From: Paul Wouters <pwouters@redhat.com>
Date: Sun, 7 Aug 2016 19:43:59 -0400
Subject: make nssdir flexible

This patch comes from upstream as a suggestion for how to break nssdir
out and treat it independently.
---
 mk/config.mk                                 |  3 +--
 programs/_import_crl/ipsec__import_crl.8.xml |  2 +-
 programs/auto/ipsec_auto.8.xml               |  4 ++--
 programs/configs/d.ipsec.conf/ipsecdir.xml   | 16 +++++-----------
 programs/ipsec/ipsec_import.8.xml            |  2 +-
 programs/ipsec/ipsec_initnss.8.xml           |  3 ++-
 programs/look/look.in                        |  4 ++--
 programs/newhostkey/ipsec_newhostkey.8.xml   |  2 +-
 programs/newhostkey/newhostkey.in            |  2 +-
 programs/pluto/ikev2_spdb_struct.c           |  9 ---------
 programs/pluto/ipsec_pluto.8.xml             |  2 +-
 programs/pluto/plutomain.c                   |  3 ++-
 programs/rsasigkey/ipsec_rsasigkey.8.xml     |  6 +++---
 programs/rsasigkey/rsasigkey.c               |  4 ++--
 programs/showhostkey/ipsec_showhostkey.8.xml |  8 ++++----
 programs/showhostkey/showhostkey.c           |  2 +-
 16 files changed, 29 insertions(+), 43 deletions(-)

diff --git a/mk/config.mk b/mk/config.mk
index 1ce7571..3d0e20a 100644
--- a/mk/config.mk
+++ b/mk/config.mk
@@ -117,7 +117,7 @@ SYSCONFDIR?=$(DESTDIR)$(FINALSYSCONFDIR)
 FINALCONFDDIR?=$(FINALCONFDIR)/ipsec.d
 CONFDDIR?=$(DESTDIR)$(FINALCONFDDIR)
 
-FINALNSSDIR?=$(FINALCONFDIR)/ipsec.d
+FINALNSSDIR?=/var/lib/ipsec
 NSSDIR?=$(DESTDIR)$(FINALNSSDIR)
 
 # sample configuration files go into
@@ -483,7 +483,6 @@ TRANSFORM_VARIABLES = sed -e "s:@IPSECVERSION@:$(IPSECVERSION):g" \
 			-e "s:@EXAMPLECONFDIR@:$(EXAMPLECONFDIR):g" \
 			-e "s:@FINALBINDIR@:$(FINALBINDIR):g" \
 			-e "s:@FINALCONFDDIR@:$(FINALCONFDDIR):g" \
-			-e "s:@FINALNSSDIR@:$(FINALNSSDIR):g" \
 			-e "s:@FINALCONFDIR@:$(FINALCONFDIR):g" \
 			-e "s:@FINALCONFFILE@:$(FINALCONFFILE):g" \
 			-e "s:@FINALDOCDIR@:$(FINALDOCDIR):g" \
diff --git a/programs/_import_crl/ipsec__import_crl.8.xml b/programs/_import_crl/ipsec__import_crl.8.xml
index 6b250db..82ab267 100644
--- a/programs/_import_crl/ipsec__import_crl.8.xml
+++ b/programs/_import_crl/ipsec__import_crl.8.xml
@@ -23,7 +23,7 @@
 <para><emphasis remap='I'>_import_crl</emphasis>
 is spawned by
 <emphasis remap='B'>pluto</emphasis>
-in order to add or update a CRL in the NSS database (default: /etc/ipsec.d)</para>
+in order to add or update a CRL in the NSS database (default: @IPSEC_NSSDIR@)</para>
 </refsect1>
 
 <refsect1 id='see_also'><title>SEE ALSO</title>
diff --git a/programs/auto/ipsec_auto.8.xml b/programs/auto/ipsec_auto.8.xml
index e6ec853..870f0d9 100644
--- a/programs/auto/ipsec_auto.8.xml
+++ b/programs/auto/ipsec_auto.8.xml
@@ -252,7 +252,7 @@ but that may change.)</para>
 </para>
 <para>      The  <option>--listcerts</option>  operation  lists  all  X.509 certificates
        loaded  locally  using  the  rightcert  and  leftcert   parameters   in
-       ipsec.conf(5). To see all certificates in the NSS database, use <option>certutil -d /etc/ipsec.d -L</option>.
+       ipsec.conf(5). To see all certificates in the NSS database, use <option>certutil -d @IPSEC_NSSDIR@ -L</option>.
 </para>
 <para>      The  <option>--checkpubkeys</option>  operation  lists  all loaded X.509 certificates
        which are about to expire or have been expired.
@@ -317,7 +317,7 @@ for details of the configuration file.
 <para>
 <literallayout remap='.nf'>
 /etc/ipsec.conf			default IPSEC configuration file
-/etc/ipsec.d/			X.509 and Opportunistic Encryption files
+@IPSEC_NSSDIR@			X.509 and Opportunistic Encryption files
 /var/run/pluto/pluto.ctl	Pluto command socket
 </literallayout>
 </para>
diff --git a/programs/configs/d.ipsec.conf/ipsecdir.xml b/programs/configs/d.ipsec.conf/ipsecdir.xml
index 954a088..b5e96a8 100644
--- a/programs/configs/d.ipsec.conf/ipsecdir.xml
+++ b/programs/configs/d.ipsec.conf/ipsecdir.xml
@@ -18,21 +18,11 @@
         <term>nsspassword</term>
         <listitem><para>
           (optional) passwords needed to unlock the NSS database
-          in /var/lib/libreswan/nss (this file should not be
+          in @IPSEC_NSSDIR@ (this file should not be
           world-readable).  See README.nss for more information.
         </para></listitem>
       </varlistentry>
       <varlistentry>
-        <term>cert9.db, key4.db, pkcs11.txt, *.db</term>
-        <listitem><para>
-          These files constitute the NSS database, which includes
-          private key material, X.509 certificates, CRLs, OCSP
-          responses, and possibly other info.  They will be created
-          and managed automatically by libreswan and should not be
-          world-readable.  See README.nss for more information.
-        </para></listitem>
-      </varlistentry>
-      <varlistentry>
         <term>policies/</term>
         <listitem><para>
           a directory containing policy group configuration
@@ -62,5 +52,9 @@
 When SELinux runs in enforced mode, changing this requires a similar
 change in the SELinux policy for the pluto daemon.
 </para>
+<para>
+  On debian systems, the NSS database itself is not controlled by this
+  value, and is instead located in @IPSEC_NSSDIR@.
+</para>
   </listitem>
   </varlistentry>
diff --git a/programs/ipsec/ipsec_import.8.xml b/programs/ipsec/ipsec_import.8.xml
index 7944a36..ee87562 100644
--- a/programs/ipsec/ipsec_import.8.xml
+++ b/programs/ipsec/ipsec_import.8.xml
@@ -27,7 +27,7 @@
 <refsect1 id='description'><title>DESCRIPTION</title>
 <para><emphasis remap='I'>ipsec import</emphasis>
 Import PKCS#12 files into the IPsec NSS database located at the
-ipsec data directory (default: /etc/ipsec.d/)
+ipsec NSS data directory (default: @IPSEC_NSSDIR@)
 </para>
 </refsect1>
 
diff --git a/programs/ipsec/ipsec_initnss.8.xml b/programs/ipsec/ipsec_initnss.8.xml
index f88a3d7..ef7b3c1 100644
--- a/programs/ipsec/ipsec_initnss.8.xml
+++ b/programs/ipsec/ipsec_initnss.8.xml
@@ -29,7 +29,8 @@
 initialises the NSS database where all private keys for RSA and
 certificate keypairs are stored. If already initialised, it will
 return an error. To remove an existing IPsec NSS database, remove all the
-*.db files and pkcs11.txt from the NSS data directory (default: /etc/ipsec.d/)
+*.db files and pkcs11.txt from the NSS data directory
+(default: @IPSEC_NSSDIR@).
 </para>
 </refsect1>
 
diff --git a/programs/look/look.in b/programs/look/look.in
index 77d7c04..231b839 100755
--- a/programs/look/look.in
+++ b/programs/look/look.in
@@ -99,12 +99,12 @@ echo ROUTING TABLES
 ip -4 route
 ip -6 route
 
-if [ -f @IPSEC_CONFDDIR@/cert9.db ]; then
+if [ -f @IPSEC_NSSDIR@/cert9.db ]; then
     echo NSS_CERTIFICATES
     certutil -L -d sql:@IPSEC_NSSDIR@ | head -4
     certutil -L -d sql:@IPSEC_NSSDIR@ | \
 	egrep -v '^NSS_CERTIFICATES|^$|Trust Attributes|MIME,' | LC_ALL=C sort
-elif [ -f @IPSEC_CONFDDIR@/cert8.db ]; then
+elif [ -f @IPSEC_NSSDIR@/cert8.db ]; then
     echo NSS_CERTIFICATES
     certutil -L -d @IPSEC_NSSDIR@ | head -4
     certutil -L -d @IPSEC_NSSDIR@ | \
diff --git a/programs/newhostkey/ipsec_newhostkey.8.xml b/programs/newhostkey/ipsec_newhostkey.8.xml
index 1229b75..d24551e 100644
--- a/programs/newhostkey/ipsec_newhostkey.8.xml
+++ b/programs/newhostkey/ipsec_newhostkey.8.xml
@@ -82,7 +82,7 @@
 	<listitem>
 	  <para>The <option>--configdir</option> option specifies the NSS
 	  configuration directory where the certificate key, and
-	  modsec databases reside (default <filename>@FINALNSSDIR@</filename>.</para>
+	  modsec databases reside (default <filename>@IPSEC_NSSDIR@</filename>.</para>
 	</listitem>
       </varlistentry>
 
diff --git a/programs/newhostkey/newhostkey.in b/programs/newhostkey/newhostkey.in
index 8e867ef..c45a90c 100755
--- a/programs/newhostkey/newhostkey.in
+++ b/programs/newhostkey/newhostkey.in
@@ -27,7 +27,7 @@ verbose=
 host=
 seeddev="--seeddev /dev/random"
 output=""
-configdir="@FINALNSSDIR@"
+configdir="@IPSEC_NSSDIR@"
 password=
 for dummy; do
     case "$1" in
diff --git a/programs/pluto/ikev2_spdb_struct.c b/programs/pluto/ikev2_spdb_struct.c
index 0d798c9..487f6e8 100644
--- a/programs/pluto/ikev2_spdb_struct.c
+++ b/programs/pluto/ikev2_spdb_struct.c
@@ -584,10 +584,6 @@ static int process_transforms(pb_stream *prop_pbs, struct print *remote_print_bu
 				passert(!sentinel_transform->valid);
 				/* save it */
 				matching_local_proposal->matching_transform[type] = sentinel_transform;
-				DBG(DBG_CONTROLMORE,
-				    DBG_log("local proposal %d type %s has %zu transforms",
-					    local_propnum, trans_type_name(type),
-					    sentinel_transform - local_transforms->transform));
 			}
 		}
 	}
@@ -728,11 +724,6 @@ static int process_transforms(pb_stream *prop_pbs, struct print *remote_print_bu
 						DBG(DBG_CONTROLMORE,
 						    struct print *buf = print_buf();
 						    print_type_transform(buf, type, &remote_transform);
-						    DBG_log("remote proposal %u transform %d (%s) matches local proposal %d type %d (%s) transform %zu",
-							    remote_propnum, remote_transform_nr,
-							    buf->buf, local_propnum,
-							    type, trans_type_name(type),
-							    local_transform - local_transforms->transform);
 						    pfree(buf));
 						*matching_local_transform = local_transform;
 						break;
diff --git a/programs/pluto/ipsec_pluto.8.xml b/programs/pluto/ipsec_pluto.8.xml
index 56e937c..2c2f4e8 100644
--- a/programs/pluto/ipsec_pluto.8.xml
+++ b/programs/pluto/ipsec_pluto.8.xml
@@ -1065,7 +1065,7 @@
 
       <para>Pluto supports X.509 Certificates. All certificate handling is done using
       the NSS library and all certificate material is stored in an NSS database in
-      /etc/ipsec.d or another directory as specified by <option>--ipsecdir</option>.</para>
+      @IPSEC_NSSDIR@ or another directory as specified by <option>--ipsecdir</option>.</para>
 
       <para>Pluto may core dump. It will normally do so into the current
       working directory. You can specify the --coredir option for pluto, or
diff --git a/programs/pluto/plutomain.c b/programs/pluto/plutomain.c
index 93629a7..a038076 100644
--- a/programs/pluto/plutomain.c
+++ b/programs/pluto/plutomain.c
@@ -1699,11 +1699,12 @@ void show_setup_plutomain(void)
 	whack_log(RC_COMMENT, "config setup options:");	/* spacer */
 	whack_log(RC_COMMENT, " ");	/* spacer */
 	whack_log(RC_COMMENT,
-		"configdir=%s, configfile=%s, secrets=%s, ipsecdir=%s, dumpdir=%s, statsbin=%s",
+		"configdir=%s, configfile=%s, secrets=%s, ipsecdir=%s, nssdir=%s, dumpdir=%s, statsbin=%s",
 		oco->confdir,
 		oco->conffile,
 		oco->secretsfile,
 		oco->confddir,
+		oco->nssdb,
 		coredir,
 		pluto_stats_binary == NULL ? "unset" :  pluto_stats_binary);
 
diff --git a/programs/rsasigkey/ipsec_rsasigkey.8.xml b/programs/rsasigkey/ipsec_rsasigkey.8.xml
index 0a1f4f8..e9703d6 100644
--- a/programs/rsasigkey/ipsec_rsasigkey.8.xml
+++ b/programs/rsasigkey/ipsec_rsasigkey.8.xml
@@ -72,7 +72,7 @@ prevented when the system is running in FIPS mode.</para>
 
 <para>The <option>--configdir</option> option specifies the nss configuration directory to use.
 This is the directory where the NSS certificate, key and security modules databases reside. The
-default value is <filename>/etc/ipsec.d</filename>.</para>
+default value is <filename>@IPSEC_NSSDIR@</filename>.</para>
 
 <para>The <option>--password</option> option specifies the nss cryptographic module authentication
 password if the NSS module has been configured to require it.  A password is required by hardware
@@ -181,8 +181,8 @@ be arbitrarily delayed if the system's entropy pool is low on
 randomness, and  the time taken by the search for primes is also somewhat
 unpredictable. Specifically, embedded systems and most virtual machines are low on
 entropy. In such a situation, consider generating the RSA key on another machine,
-and copying <filename>ipsec.secrets</filename> and the <filename>ipsec.d/*db</filename>
-files to the embedded platform. Note that NSS embeds the full path in the DB files, so
+and copying <filename>ipsec.secrets</filename> and the <filename>@IPSEC_NSSDIR@</filename>
+directory tree to the embedded platform. Note that NSS embeds the full path in the DB files, so
 the path on proxy machine must be identical to the path on the destination machine.
 </para>
 </refsect1>
diff --git a/programs/rsasigkey/rsasigkey.c b/programs/rsasigkey/rsasigkey.c
index 2d075ae..365a0ae 100644
--- a/programs/rsasigkey/rsasigkey.c
+++ b/programs/rsasigkey/rsasigkey.c
@@ -201,7 +201,7 @@ int main(int argc, char *argv[])
 			break;
 		case 'c':       /* nss configuration directory */
 		case 'd':       /* -d is used for configdir with nss tools */
-			lsw_conf_confddir(optarg);
+			lsw_conf_nssdb(optarg);
 			break;
 		case 'P':       /* token authentication password */
 			lsw_conf_nsspassword(optarg);
@@ -287,7 +287,7 @@ void rsasigkey(int nbits, int seedbits, const struct lsw_conf_options *oco)
 	realtime_t now = realnow();
 
 	lsw_nss_buf_t err;
-	if (!lsw_nss_setup(oco->confddir, 0, lsw_nss_get_password, err)) {
+	if (!lsw_nss_setup(oco->nssdb, 0, lsw_nss_get_password, err)) {
 		fprintf(stderr, "%s: %s\n", progname, err);
 		exit(1);
 	}
diff --git a/programs/showhostkey/ipsec_showhostkey.8.xml b/programs/showhostkey/ipsec_showhostkey.8.xml
index 4184dab..7e441be 100644
--- a/programs/showhostkey/ipsec_showhostkey.8.xml
+++ b/programs/showhostkey/ipsec_showhostkey.8.xml
@@ -105,7 +105,7 @@
 	  <para>
 	    Specify the libreswan configuration directory that
 	    contains the NSS database (default
-	    <filename>@FINALNSSDIR@</filename>).
+	    <filename>@IPSEC_NSSDIR@</filename>).
 	  </para>
 	</listitem>
       </varlistentry>
@@ -118,7 +118,7 @@
 	  <para>
 	    Specify the password to use when accessing the NSS
 	    database (default contained in
-	    <filename>@FINALNSSDIR@/nsspassword</filename>).
+	    <filename>@IPSEC_NSSDIR@/nsspassword</filename>).
 	  </para>
 	</listitem>
       </varlistentry>
@@ -272,8 +272,8 @@ IN    IPSECKEY  10 1 2 10.11.12.13  AQOF8tZ2...+buFuFn/"
 <refsect1 id='files'>
   <title>FILES</title>
   <para>
-    <filename>@FINALNSSDIR@</filename>,
-    <filename>@FINALNSSDIR@/nsspassword</filename>
+    <filename>@IPSEC_NSSDIR@</filename>,
+    <filename>@IPSEC_NSSDIR@/nsspassword</filename>
   </para>
 </refsect1>
 
diff --git a/programs/showhostkey/showhostkey.c b/programs/showhostkey/showhostkey.c
index 1341318..de1a9f8 100644
--- a/programs/showhostkey/showhostkey.c
+++ b/programs/showhostkey/showhostkey.c
@@ -453,7 +453,7 @@ int main(int argc, char *argv[])
 	 * processed, and really are "constant".
 	 */
 	const struct lsw_conf_options *oco = lsw_init_options();
-	libreswan_log("using config directory \"%s\"\n", oco->confddir);
+	libreswan_log("using config directory \"%s\"\n", oco->nssdb);
 
 	/*
 	 * Set up for NSS - contains key pairs.
